### PR TITLE
Improve rank mismatch errors in zipperings between ranges / default arrays

### PR DESCRIPTION
**Pull request by bradcray**
In a post-post-meeting question this week, @stonea asked about zippering
things of different ranks.  A quick check showed that for common types,
we generated an error, but not a very clear one, related more to the underlying
implementations than anything a user should be seeing.  This improves that error
for ranges and default domains and arrays by doing a simple rank check in the
follower iterators.

TBD: Do we need to do additional work to support the same checks in distributed
domains and arrays, or will the fact that they are implemented in terms of local
domains and arrays cause this check to be inherited?  I'm not certain and need to
write some more tests for that case (in a separate effort).

### CODE REVIEW

File: modules/internal/ChapelRange.chpl [hunk 1]
```diff
@@ -2455,7 +2455,8 @@ operator :(r: range(?), type t: range(?)) {
       compilerError("iteration over a range with no first index");
 
     if followThis.size != 1 then
-      compilerError("iteration over a range with multi-dimensional iterator");
+      compilerError("rank mismatch in zippered iteration (leader is " +
+                    followThis.size:string + "D, but ranges are 1D)");
```
Comment (vasslitvinov) [inline modules/internal/ChapelRange.chpl:2459]: This is surely an improvement. I suggest clarifying it further by using the word "follower", ex.
`(leader is 3D, but follower is a range, which is 1D)`

File: modules/internal/DefaultRectangular.chpl [hunk 2]
```diff
@@ -489,6 +489,11 @@ module DefaultRectangular {
                offset=createTuple(rank, intIdxType, 0:intIdxType))
       where tag == iterKind.follower {
 
+      if followThis.size != this.rank then
+        compilerError("rank mismatch in zippered iteration (leader is " +
+                      followThis.size:string + "D, but array is " +
+                      this.rank:string + "D)");
```
Comment (vasslitvinov) [inline modules/internal/DefaultRectangular.chpl:495]: Again, I'd like to see the word "follower" to accompany "leader", ex.
`... but array follower is 2D`

**Review by vasslitvinov — APPROVED** [review]
Looks good, assuming the minor improvement to the error messages I suggest.

File: modules/internal/ChapelRange.chpl (hunk 1 above)
Comment (bradcray) [reply modules/internal/ChapelRange.chpl:2459]: (implied, but I should've said it:  So let me know whether you agree with the rationale for the current message, still prefer your original proposal, prefer something else, or would like me to create more of a design discussion around this).

**Comment by vasslitvinov** [comment]
@bradcray feel free to stick with your wording, although not my preference. Or consider the alternative you suggest:

`rank mismatch in zippered iteration (a 2D expression is [being] zipped with a range, which is 1D)`
`rank mismatch in zippered iteration (a 2D expression is [being] zipped with a 1D array)`

**Pull request merged**